appcues/src/main/java/com/appcues/analytics/ExperienceLifecycleTracker.kt

```diff
@@ -16,73 +17,61 @@ import com.appcues.statemachine.State.EndingExperience
-import com.appcues.statemachine.StateMachine
-import kotlinx.coroutines.CoroutineDispatcher
-import kotlinx.coroutines.Dispatchers
-import kotlinx.coroutines.launch
-import kotlinx.coroutines.withContext
 import org.koin.core.component.KoinScopeComponent
 import org.koin.core.component.inject
 import org.koin.core.scope.Scope
 import java.util.Date
 
 internal class ExperienceLifecycleTracker(
-    override val scope: Scope
+    override val scope: Scope,
 ) : KoinScopeComponent {
 
-    // lazy property injection to avoid circular DI reference in constructor
-    // AnalyticsTracker -> this <- Analytics Tracker
+    // Circular dependency AnalyticsTracker -> Processor -> ExperienceRenderer -> this
     private val analyticsTracker: AnalyticsTracker by inject()
-    private val stateMachine: StateMachine by inject()
     private val storage: Storage by inject()
+    private val config: AppcuesConfig by inject()
 
-    suspend fun start(dispatcher: CoroutineDispatcher = Dispatchers.IO): Unit = withContext(dispatcher) {
-        launch { observeState() }
-        launch { observeErrors() }
-    }
+    fun onState(state: State): Unit = with(state) {
+        if (shouldNotTrackUnpublished()) return
 
-    private suspend fun observeState() {
-        stateMachine.stateFlow.collect {
-            if (it.shouldTrack()) { // will not track for unpublished (preview) experiences
-                when (it) {
-                    is RenderingStep -> {
-                        if (it.isFirst) {
-                            // update this value for auto-properties
-                            storage.lastContentShownAt = Date()
-                            trackLifecycleEvent(ExperienceStarted(it.experience), SdkMetrics.trackRender(it.experience.requestId))
-                        }
-                        trackLifecycleEvent(StepSeen(it.experience, it.flatStepIndex))
-                    }
-                    is EndingStep -> {
-                        if (it.markComplete) {
-                            trackLifecycleEvent(StepCompleted(it.experience, it.flatStepIndex))
-                        }
-                    }
-                    is EndingExperience -> {
-                        if (it.markComplete) {
-                            // if ending on the last step OR an action requested it be considered complete explicitly,
-                            // track the experience_completed event
-                            trackLifecycleEvent(ExperienceCompleted(it.experience))
-                        } else {
-                            // otherwise its considered experience_dismissed (not completed)
-                            trackLifecycleEvent(ExperienceDismissed(it.experience, it.flatStepIndex))
-                        }
-                    }
-                    else -> Unit
+        when (this) {
+            is BeginningExperience -> config.experienceListener?.experienceStarted(experience.id)
```

Comment:
ok so this is replacing what used to be in the state machine logic, trying to follow along.

Reply:
I was trying to move stuff out of the statemachine and leave if with as less dependencies as possible, since this is related to the experience lifecycle i was hoping we could plug it in here